kafka: return partition error metadata if no term

andrwng opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->
When the metadata cache is missing term information for a given partition (e.g.
when starting up) we return an invalid (-1) term with no error.

The Java client treats this metadata as unreliable and forgets about any cached
epochs, regardless of any errors[1].

Franz go skips processing the partition altogether if there is an error,
regardless of whether there is a missing term, opting to retry later. It
otherwise accepts the missing term (-1) as a valid term and treating
it as a term moving backwards[2].

This updates the metadata handler to return an error, which seems like it
shouldn't affect the Java client, but should help Franz go ignore the invalid
metadata.

Relevant Redpanda commits in reverse chronological order:
- https://github.com/redpanda-data/redpanda/commit/a441290876ff9bb01d6d58532f388b94f3719bfe
- https://github.com/redpanda-data/redpanda/commit/7a60b758ad9ff9f2a2d7ec343f2a404ffaf51e89
- https://github.com/redpanda-data/redpanda/commit/86d583b486d033377be4c9f2b331be25ff6d9632

1. https://github.com/apache/kafka/blob/9529003fffd93a9d7e3f6ff7ab081ed84942fd13/clients/src/main/java/org/apache/kafka/clients/Metadata.java#L596-L601
2. https://github.com/twmb/franz-go/blob/3affad808a82ebfe1555d01d27732431093ac8e1/pkg/kgo/metadata.go#L811-L857


## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* None
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

andrwng commented:
/ci-repeat 3
skip-redpanda-build
skip-units
dt-repeat=5
tests/rptest/tests/write_caching_fi_e2e_test.py::WriteCachingFailureInjectionE2ETest.test_crash_all

andrwng commented:
/ci-repeat 3
skip-redpanda-build
skip-units
dt-repeat=5
rp_repo=nightly
rp_version=latest
tests/rptest/tests/write_caching_fi_e2e_test.py::WriteCachingFailureInjectionE2ETest.test_crash_all

andrwng commented:
/ci-repeat 1

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--db2164ee-f9ad-4f4b-b96e-e9ae2b6529a1-->
<!--TEST RESULTS-->
<!--6a44f71b-337f-4ae6-8dd0-1b23670a5559-->
<!--TEST RESULTS-->
<!--12abd95d-55ca-4774-9841-8af97a8bd6f2-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76392">build#76392</a></summary>


| test_class                 | test_method                                  | test_arguments                                                             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                             |
|:---------------------------|:---------------------------------------------|:---------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeThrottlingTest     | test_backlog_metric                          | {"catalog_type": "nessie", "cloud_storage_type": 1}                        | integration | https://buildkite.com/redpanda/redpanda/builds/76392#019a848e-c9fb-4737-85c7-bd9512a002c7 | FAIL          | 0/1      |                                                                                                                                                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeThrottlingTest&test_method=test_backlog_metric                 |
| EndToEndShadowIndexingTest | test_reset_spillover                         | {"cloud_storage_type": 2}                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/76392#019a848e-c9fb-4737-85c7-bd9512a002c7 | FAIL          | 0/1      |                                                                                                                                                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndShadowIndexingTest&test_method=test_reset_spillover            |
| SegmentMsTest              | test_segment_rolling_with_retention_consumer | null                                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/76392#019a848e-c9ff-4e9b-a482-bf47b75f9b56 | FLAKY         | 19/21    | upstream reliability is '92.63311451495258'. current run reliability is '90.47619047619048'. drift is 2.15692 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
| SIPartitionMovementTest    | test_cross_shard                             | {"cloud_storage_type": 1, "num_to_upgrade": 0, "with_cloud_topics": false} | integration | https://buildkite.com/redpanda/redpanda/builds/76392#019a848e-c9fb-4737-85c7-bd9512a002c7 | FAIL          | 0/1      |                                                                                                                                                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SIPartitionMovementTest&test_method=test_cross_shard                   |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76402">build#76402</a></summary>


| test_class                          | test_method    | test_arguments              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                     |
|:------------------------------------|:---------------|:----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| WriteCachingFailureInjectionE2ETest | test_crash_all | {"use_transactions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/76402#019a8513-6ae9-4b1b-aa67-2bce58ec5a73 | FLAKY         | 21/25    | upstream reliability is '91.24613800205972'. current run reliability is '80.95238095238095'. drift is 10.29376 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76578">build#76578</a></summary>


| test_class                              | test_method                                                     | test_arguments                                           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                                       |
|:----------------------------------------|:----------------------------------------------------------------|:---------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterLinkingTopicSyncingWithTlsFiles  | test_topic_syncing_include_exclude                              | null                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9861-a272-41b3-af43-2e385eddce87 | FLAKY         | 16/21    | upstream reliability is '100.0'. current run reliability is '76.19047619047619'. drift is 23.80952 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterLinkingTopicSyncingWithTlsFiles&test_method=test_topic_syncing_include_exclude            |
| ClusterLinkingTopicSyncingWithTlsValues | test_topic_syncing_include_exclude                              | null                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9861-a276-4899-b78b-c51067db0427 | FLAKY         | 17/21    | upstream reliability is '99.76470588235294'. current run reliability is '80.95238095238095'. drift is 18.81232 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterLinkingTopicSyncingWithTlsValues&test_method=test_topic_syncing_include_exclude           |
| ClusterRateQuotaTest                    | test_client_group_consume_rate_throttle_mechanism               | null                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9861-a274-4286-97fd-577836d1d262 | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_consume_rate_throttle_mechanism               |
| ClusterRateQuotaTest                    | test_client_group_produce_rate_throttle_mechanism               | null                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9864-f017-41c8-8c7a-25ac7b9b71b8 | FLAKY         | 12/21    | upstream reliability is '100.0'. current run reliability is '57.14285714285714'. drift is 42.85714 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_produce_rate_throttle_mechanism               |
| ClusterRateQuotaTest                    | test_client_response_throttle_mechanism_applies_to_next_request | null                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9861-a27b-413d-8266-36a917601be2 | FLAKY         | 16/21    | upstream reliability is '100.0'. current run reliability is '76.19047619047619'. drift is 23.80952 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism_applies_to_next_request |
| ClusterRateQuotaTest                    | test_client_response_throttle_mechanism_applies_to_next_request | null                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9864-f01c-4df8-92b0-8e098ec60a86 | FLAKY         | 13/21    | upstream reliability is '100.0'. current run reliability is '61.904761904761905'. drift is 38.09524 and the allowed drift is set to 50. The test should PASS            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism_applies_to_next_request |
| DatalakeClusterRestoreTest              | test_basic                                                      | {"catalog_type": "nessie", "cloud_storage_type": 1}      | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9864-f017-41c8-8c7a-25ac7b9b71b8 | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_basic                                                |
| DatalakeClusterRestoreTest              | test_basic                                                      | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9861-a276-4899-b78b-c51067db0427 | FLAKY         | 5/21     | upstream reliability is '100.0'. current run reliability is '23.809523809523807'. drift is 76.19048 and the allowed drift is set to 50. The test should FAIL            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_basic                                                |
| DatalakeClusterRestoreTest              | test_basic                                                      | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9864-f018-40ef-9c70-28247ed8971c | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_basic                                                |
| DatalakeClusterRestoreTest              | test_basic                                                      | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}   | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9861-a277-4d16-8e50-ad5fc29eabf8 | FLAKY         | 4/21     | upstream reliability is '100.0'. current run reliability is '19.047619047619047'. drift is 80.95238 and the allowed drift is set to 50. The test should FAIL            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_basic                                                |
| DatalakeClusterRestoreTest              | test_restore_partition_spec                                     | {"catalog_type": "nessie", "cloud_storage_type": 1}      | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9861-a279-45cc-823f-c48e56b780b4 | FLAKY         | 4/21     | upstream reliability is '100.0'. current run reliability is '19.047619047619047'. drift is 80.95238 and the allowed drift is set to 50. The test should FAIL            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_restore_partition_spec                               |
| DatalakeClusterRestoreTest              | test_restore_partition_spec                                     | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}   | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9861-a27c-4f39-b669-ac7b0b92ae4b | FLAKY         | 5/21     | upstream reliability is '100.0'. current run reliability is '23.809523809523807'. drift is 76.19048 and the allowed drift is set to 50. The test should FAIL            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_restore_partition_spec                               |
| DatalakeClusterRestoreTest              | test_slow_tiered_storage_dlq                                    | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9861-a271-4dcb-9ca9-665cb226b5f2 | FLAKY         | 3/21     | upstream reliability is '100.0'. current run reliability is '14.285714285714285'. drift is 85.71429 and the allowed drift is set to 50. The test should FAIL            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_slow_tiered_storage_dlq                              |
| DatalakeClusterRestoreTest              | test_slow_tiered_storage_dlq                                    | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}   | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9861-a272-41b3-af43-2e385eddce87 | FLAKY         | 8/21     | upstream reliability is '100.0'. current run reliability is '38.095238095238095'. drift is 61.90476 and the allowed drift is set to 50. The test should FAIL            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_slow_tiered_storage_dlq                              |
| DatalakeClusterRestoreTest              | test_slow_tiered_storage_dupe_records                           | {"catalog_type": "nessie", "cloud_storage_type": 1}      | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9861-a274-4286-97fd-577836d1d262 | FLAKY         | 1/21     | upstream reliability is '100.0'. current run reliability is '4.761904761904762'. drift is 95.2381 and the allowed drift is set to 50. The test should FAIL              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_slow_tiered_storage_dupe_records                     |
| DatalakeClusterRestoreTest              | test_slow_tiered_storage_dupe_records                           | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9861-a275-48e4-9234-553060562d4c | FLAKY         | 6/21     | upstream reliability is '100.0'. current run reliability is '28.57142857142857'. drift is 71.42857 and the allowed drift is set to 50. The test should FAIL             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_slow_tiered_storage_dupe_records                     |
| DatalakeClusterRestoreTest              | test_slow_tiered_storage_dupe_records                           | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}   | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9861-a276-4899-b78b-c51067db0427 | FLAKY         | 3/21     | upstream reliability is '100.0'. current run reliability is '14.285714285714285'. drift is 85.71429 and the allowed drift is set to 50. The test should FAIL            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_slow_tiered_storage_dupe_records                     |
| MountUnmountIcebergTest                 | test_simple_remount                                             | {"cloud_storage_type": 1}                                | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9864-f014-4931-991e-e15146fe4536 | FAIL          | 0/5      | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                                          |
| LogCompactionTxRemovalUpgradeTest       | test_tx_control_batch_removal_with_upgrade                      | {"test_case_name": "All commits"}                        | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9861-a279-45cc-823f-c48e56b780b4 | FLAKY         | 9/21     | upstream reliability is '85.9375'. current run reliability is '42.857142857142854'. drift is 43.08036 and the allowed drift is set to 50. The test should PASS          | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalUpgradeTest&test_method=test_tx_control_batch_removal_with_upgrade         |
| LogCompactionTxRemovalUpgradeTest       | test_tx_control_batch_removal_with_upgrade                      | {"test_case_name": "Mixed aborts and commits"}           | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9861-a27b-413d-8266-36a917601be2 | FLAKY         | 15/21    | upstream reliability is '90.94202898550725'. current run reliability is '71.42857142857143'. drift is 19.51346 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalUpgradeTest&test_method=test_tx_control_batch_removal_with_upgrade         |
| SegmentMsTest                           | test_segment_rolling_with_retention_consumer                    | null                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9861-a274-4286-97fd-577836d1d262 | FLAKY         | 19/21    | upstream reliability is '92.08633093525181'. current run reliability is '90.47619047619048'. drift is 1.61014 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer                           |
| MultiRestartTest                        | test_recovery_after_multiple_restarts                           | {"cloud_storage_type": 2}                                | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9864-f017-41c8-8c7a-25ac7b9b71b8 | FLAKY         | 20/21    | upstream reliability is '99.59839357429718'. current run reliability is '95.23809523809523'. drift is 4.3603 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MultiRestartTest&test_method=test_recovery_after_multiple_restarts                               |
| OffsetForLeaderEpochReadReplicaTest     | test_offset_for_leader_epoch                                    | null                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9864-f016-4d2b-8d25-fd4b579fda31 | FLAKY         | 15/21    | upstream reliability is '100.0'. current run reliability is '71.42857142857143'. drift is 28.57143 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=OffsetForLeaderEpochReadReplicaTest&test_method=test_offset_for_leader_epoch                     |
| RpkRegistryTest                         | test_produce_consume_json                                       | null                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9864-f014-4931-991e-e15146fe4536 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RpkRegistryTest&test_method=test_produce_consume_json                                            |
| TxUpgradeCompactionTest                 | upgrade_with_compaction_test                                    | null                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/76578#019a9864-f015-4a5b-aa93-0ccc258f239d | FLAKY         | 20/21    | upstream reliability is '97.35099337748345'. current run reliability is '95.23809523809523'. drift is 2.1129 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TxUpgradeCompactionTest&test_method=upgrade_with_compaction_test                                 |
</details>
<!--/body-->

andrwng commented:
/ci-repeat 3
skip-redpanda-build
skip-units
dt-repeat=5
tests/rptest/tests/write_caching_fi_e2e_test.py::WriteCachingFailureInjectionE2ETest.test_crash_all

andrwng committed:
kafka: return partition error metadata if no term

When the metadata cache is missing term information for a given partition (e.g.
when starting up) we return an invalid (-1) term with no error.

The Java client treats this metadata as unreliable and forgets about any cached
epochs, regardless of any errors[1].

Franz go skips processing the partition altogether if there is an error,
regardless of whether there is a missing term, opting to retry later. It
otherwise accepts the missing term (-1) as a valid term and treating
it as a term moving backwards[2].

This updates the metadata handler to return an error, which seems like it
shouldn't affect the Java client, but should help Franz go ignore the invalid
metadata.

Relevant Redpanda commits in reverse chronological order:
- https://github.com/redpanda-data/redpanda/commit/a441290876ff9bb01d6d58532f388b94f3719bfe
- https://github.com/redpanda-data/redpanda/commit/7a60b758ad9ff9f2a2d7ec343f2a404ffaf51e89
- https://github.com/redpanda-data/redpanda/commit/86d583b486d033377be4c9f2b331be25ff6d9632

1. https://github.com/apache/kafka/blob/9529003fffd93a9d7e3f6ff7ab081ed84942fd13/clients/src/main/java/org/apache/kafka/clients/Metadata.java#L596-L601
2. https://github.com/twmb/franz-go/blob/3affad808a82ebfe1555d01d27732431093ac8e1/pkg/kgo/metadata.go#L811-L857

andrwng commented:
/ci-repeat 3
skip-redpanda-build
skip-units
dt-repeat=5
tests/rptest/tests/write_caching_fi_e2e_test.py::WriteCachingFailureInjectionE2ETest.test_crash_all

andrwng commented:
/ci-repeat 6
skip-redpanda-build
skip-units
dt-repeat=10
tests/rptest/tests/write_caching_fi_e2e_test.py::WriteCachingFailureInjectionE2ETest.test_crash_all

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/76578-->
<!--8c43a9bb-2955-45ed-919c-2896cd28c37b-->
<!--8e0e48d4-c82f-48e0-a239-29abcdc8ee77-->
<!--f56981be-df10-49bb-8bc6-3b50e24645bd-->
<!--1e36fad3-c3c1-41ea-ad11-05282bfe7ade-->
<!--f24e981e-357d-46b3-aa54-cbc7df23e929-->
<!--70ef4e00-8772-4439-baea-c148e3dfb6ae-->
<!--61351f45-e7ac-473d-a506-5b6f7c81e4b1-->
<!--b95e89f4-0228-4cce-bf70-81eb780d10e1-->
<!--bc84d45f-4537-4531-b356-b4ba5168a12a-->
<!--title-->
## Retry command for [Build#76578](https://buildkite.com/redpanda/redpanda/builds/76578)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records@{"catalog_type":"rest_hadoop","cloud_storage_type":1}
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_restore_partition_spec@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq@{"catalog_type":"rest_hadoop","cloud_storage_type":1}
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_basic@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records@{"catalog_type":"nessie","cloud_storage_type":1}
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_restore_partition_spec@{"catalog_type":"nessie","cloud_storage_type":1}
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_basic@{"catalog_type":"rest_hadoop","cloud_storage_type":1}
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
tests/rptest/tests/datalake/mount_unmount_test.py::MountUnmountIcebergTest.test_simple_remount@{"cloud_storage_type":1}
tests/rptest/tests/rpk_registry_test.py::RpkRegistryTest.test_produce_consume_json
```
<!--/body-->

WillemKauf commented:
@andrwng can we revisit this?

andrwng commented:
>will this fix also fix -1 returned during leadership transitions? If so then maybe it would fix the same flake I'm trying to fix here: https://github.com/redpanda-data/redpanda/pull/30654

Yeah, looking through your description, it's addressing the same write caching test that this is trying to fix. In that test scenario, we would return epoch 0 and an error code, and franz go has recently been changed to update so it will retry in that scenario.

I'll close this draft since the fix ended up pulling in some other change, but I'm retesting on this PR https://github.com/redpanda-data/redpanda/pull/30538
